comms/src/connection_manager/manager.rs

```diff
@@ -142,3 +174,113 @@ impl ConnectionManager {
+            consumer_address,
+            socks_proxy_address: None,
+        }
+    }
+
+    fn make_peer_manager(peers: Vec<Peer<CommsPublicKey>>) -> Arc<PeerManager<CommsPublicKey, CommsDataStore>> {
+        Arc::new(factories::peer_manager::create().with_peers(peers).build().unwrap())
+    }
+
+    #[test]
+    #[allow(non_snake_case)]
+    fn establish_peer_connection_by_peer() {
+        log::init();
+        let context = Context::new();
+
+        let dispatcher = Dispatcher::new(handlers::ControlServiceResolver::new()).route(
+            ControlServiceMessageType::EstablishConnection,
+            handlers::establish_connection,
+        );
+
+        //---------------------------------- Node B Setup --------------------------------------------//
+
+        let node_B_consumer_address = InprocAddress::random();
+        let node_B_control_port_address: NetAddress = "127.0.0.1:9000".parse().unwrap();
+
+        let node_B_msg_counter = ConnectionMessageCounter::new(&context);
+        node_B_msg_counter.start(node_B_consumer_address.clone());
+
+        let node_B_peer = factories::peer::create()
+            .with_net_addresses(vec![node_B_control_port_address.clone()])
+            .build()
+            .unwrap();
+
+        // Node B knows no peers
+        let node_B_peer_manager = make_peer_manager(vec![]);
+        let node_B_connection_manager = Arc::new(ConnectionManager::new(
+            node_B_peer_manager,
+            make_peer_connection_config(&context, node_B_consumer_address.clone()),
+        ));
+
+        // Start node B's control service
+        let node_B_control_service = ControlService::new(&context)
+            .configure(ControlServiceConfig {
+                socks_proxy_address: None,
+                listener_address: node_B_control_port_address,
+            })
+            .serve(dispatcher, node_B_connection_manager)
+            .unwrap();
+
+        //---------------------------------- Node A setup --------------------------------------------//
+
+        let node_A_consumer_address = InprocAddress::random();
+
+        // Add node B to node A's peer manager
+        let node_A_peer_manager = make_peer_manager(vec![node_B_peer.clone()]);
+        let node_A_connection_manager = Arc::new(ConnectionManager::new(
+            node_A_peer_manager,
+            make_peer_connection_config(&context, node_A_consumer_address),
+        ));
+
+        //------------------------------ Negotiate connection to node B -----------------------------------//
+
+        let to_node_B_conn = node_A_connection_manager
+            .establish_connection_to_peer(&node_B_peer)
+            .unwrap();
+
+        to_node_B_conn.set_linger(Linger::Indefinitely).unwrap();
+
+        // Wait for a second for the accept message to be sent from the control service handler
+        // TODO: This connection should be properly accepted before being returned
+        std::thread::sleep(std::time::Duration::from_millis(1000));
```

Comment:
Some of the modules are using the pause function to create a delay: `    fn pause() {
        thread::sleep(Duration::from_millis(5));
    }`

Reply:
I don't want to keep this - this is so that the test waits a bit so that node B's control service can send the accept message, I plan to fix this in future - so I don't want this to look permanent :)